components/butterfly/src/network.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2018 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The Butterfly network abstraction.
+//!
+//! The abstraction provides communication channels for sending SWIM
+//! and gossip messages.
+
+use std::fmt::Debug;
+use std::marker::{Send, Sized};
+use std::net::SocketAddr;
+
+use error::Result;
+
+/// A trait for types that can fail when cloning. Mostly meant for
+/// things like communication channels which usually use some
+/// operating system resources and because of it, cloning/duplicating
+/// them may fail for whatever reason.
+pub trait TryClone: Sized {
+    fn try_clone(&self) -> Result<Self>;
+}
+
+// TODO(krnowak): See a TODO about Debug for Network trait below.
+/// A trait for types used for communicating with SWIM messages.
+pub trait SwimChannel: TryClone + Send + Debug {
+    /// Send a SWIM message (as bytes) to the given address. The
+    /// returned value holds a number of bytes sent.
+    fn send(&self, buf: &[u8], addr: SocketAddr) -> Result<usize>;
+    /// Receive a SWIM message (as bytes) from the channel. The
+    /// returned value holds the size and an address from where the
+    /// bytes came.
+    fn receive(&self, buf: &mut [u8]) -> Result<(usize, SocketAddr)>;
+}
+
+/// A trait for types used for sending gossip messages (rumors).
+pub trait GossipSender {
+    /// Send a rumor (as bytes).
+    fn send(&self, buf: &[u8]) -> Result<()>;
+}
+
+/// A trait for types used for receiving gossip messages (rumors).
+pub trait GossipReceiver {
+    /// Receive a rumor (as bytes).
+    fn receive(&self) -> Result<Vec<u8>>;
+}
+
+// TODO(krnowak): Not sure if this static lifetime specifier here is a
+// correct thing to do. It is either here on in several other places
+// where generic type N constrained to being an implementation of the
+// Network trait is used (trace, expire, inbound, outbound and so
+// on). I added it here, because Network is exclusively used by the
+// butterfly component.
+//
+// Same for Debug - Network is used by the butterfly component only so
+// I add it here to save me some typing.
+/// A trait for types used to provide SWIM and gossip communication
+/// channels.
+pub trait Network: Send + Sync + Debug + 'static {
```

Comment:
I was basically getting these errors:

```
error[E0310]: the parameter type `N` may not live long enough
   --> src/server/mod.rs:252:14
    |
126 | impl<N: Network> Server<N> {
    |      -- help: consider adding an explicit lifetime bound `N: 'static`...
...
252 |             .spawn(move || {
    |              ^^^^^
    |
note: ...so that the type `[closure@src/server/mod.rs:252:20: 255:14 server_a:server::Server<N>, swim_receiver:<N as network::Network>::SwimReceiver, swim_sender_a:<N as network::Network>::SwimSender, tx_outbound:std::sync::mpsc::Sender<(std::net::SocketAddr, message::swim::Swim)>]` will meet its required lifetime bounds
   --> src/server/mod.rs:252:14
    |
252 |             .spawn(move || {
    |              ^^^^^

error[E0310]: the associated type `<N as network::Network>::SwimReceiver` may not live long enough
   --> src/server/mod.rs:252:14
    |
252 |             .spawn(move || {
    |              ^^^^^
    |
    = help: consider adding an explicit lifetime bound `<N as network::Network>::SwimReceiver: 'static`...
note: ...so that the type `[closure@src/server/mod.rs:252:20: 255:14 server_a:server::Server<N>, swim_receiver:<N as network::Network>::SwimReceiver, swim_sender_a:<N as network::Network>::SwimSender, tx_outbound:std::sync::mpsc::Sender<(std::net::SocketAddr, message::swim::Swim)>]` will meet its required lifetime bounds
   --> src/server/mod.rs:252:14
    |
252 |             .spawn(move || {
    |              ^^^^^
```

So I either would have to add this static lifetime in every place I use the `Network` trait or its associated types or add it just for the `Network` trait. I guess the reason is that `Network` trait is used in the detached thread that lives as long as the process itself.